tokio/tests/io_mem_stream.rs

```diff
@@ -62,20 +64,47 @@ async fn disconnect() {
     t2.await.unwrap();
 }
 
+#[tokio::test]
+async fn disconnect_reader() {
+    let (a, mut b) = duplex(2);
+
+    let t1 = tokio::spawn(async move {
+        // this will block, as not all data fits into duplex
+        b.write_all(b"ping").await.unwrap_err();
+    });
+
+    let t2 = tokio::spawn(async move {
+        // here we drop the reader side, and we expect the writer in the other
+        // task to exit with an error
+        let _moved = a;
+    });
+
+    // wait for drop first
+    t2.await.unwrap();
+    // then try and join writer task
+    t1.await.unwrap();
+}
+
 #[tokio::test]
 async fn max_write_size() {
     let (mut a, mut b) = duplex(32);
 
+    // needs a barrier to avoid droping b before we can perform the second write
+    let r_barrier = Arc::new(Notify::new());
+    let w_barrier = r_barrier.clone();
+
     let t1 = tokio::spawn(async move {
         let n = a.write(&[0u8; 64]).await.unwrap();
         assert_eq!(n, 32);
         let n = a.write(&[0u8; 64]).await.unwrap();
         assert_eq!(n, 4);
+        w_barrier.notify_waiters();
     });
 
     let t2 = tokio::spawn(async move {
         let mut buf = [0u8; 4];
         b.read_exact(&mut buf).await.unwrap();
+        r_barrier.notified().await;
     });
 
     t1.await.unwrap();
```

Comment:
You don't need the barrier if you run `t2` directly in the task.
```suggestion
    let mut buf = [0u8; 4];
    b.read_exact(&mut buf).await.unwrap();
    r_barrier.notified().await;

    t1.await.unwrap();
    // Drop b after t1 finishes writing.
    drop(b);
```